### PR TITLE
Unify Scala and Java Codegen Inputs

### PR DESCRIPTION
**Pull request by leo-da**
- Extract `language-support/codegen-common`
- Align Scala Codegen Main argument with the Java Codegen. Use option parser from the `language-support/codegen-common` (there is one important difference, Scala Codegen does not allow mapping dars to different package names, all dars have to be mapped to the same package name)
- Replace Scala Codegen println's with scala logging, to respect  configured codegen verbosity

### TODO

- [x] Validate manually that release works with `codegen-common`

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [x] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Add a line to the [release notes](https://github.com/digital-asset/daml/blob/master/docs/source/support/release-notes.rst), if appropriate

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

**Comment by leo-da** [comment]
> Please also create a shaded_binary for the combined codegen executable like we do for the [java codegen](https://github.com/digital-asset/daml/pull/585/files#diff-9d1917d721d6a5202c52ff7759dd6854R93).
> Do we want to include the combined codegen already into the SDK?

if we are to include the combined condegen front-end into the SDK, this should be done as a separate story. This PR is already getting too big.

File: language-support/scala/codegen/src/main/scala/com/digitalasset/codegen/Main.scala [hunk 1]
```diff
@@ -4,47 +4,54 @@
...
-      case Some(config) =>
-        CodeGen.generateCode(
-          config.inputFiles.toList,
-          config.packageName,
-          config.outputDir,
-          config.codeGenMode)
-      case None => // arguments are bad, error message will have been displayed
+  private def logUnsupportedEventDecoderOverride(mapping: Option[(String, String)]): Unit =
+    mapping.foreach { decoderMapping =>
+      logger.warn(s"$codegenId does not allow overriding Event Decoder, requested: $decoderMapping")
```
Comment (S11001001) [inline language-support/scala/codegen/src/main/scala/com/digitalasset/codegen/Main.scala]:
```suggestion
      logger.warn(s"$codegenId does not allow overriding Event Decoder, requested: ${decoderMapping: String}")
```

Which will probably not type-check, suggesting a different formatting of the message...?

File: language-support/scala/codegen/src/main/scala/com/digitalasset/codegen/Main.scala [hunk 2]
```diff
@@ -4,47 +4,54 @@
...
+    }
+
+  private def darsAndOnePackageName(darMap: Map[Path, Option[String]]): (List[File], String) = {
+    val empty: (List[File], Set[String]) = (List.empty, Set.empty)
+    val (dars, packageNames) =
+      darMap.foldRight(empty)((a, as) => (a._1.toFile :: as._1, as._2 + a._2.getOrElse("daml")))
```
Comment (S11001001) [inline language-support/scala/codegen/src/main/scala/com/digitalasset/codegen/Main.scala]: You should calculate `dars` and `packageNames` separately from `darMap.keys` and `darMap.values` respectively; no explicit fold necessary.

File: language-support/java/codegen/src/test/scala/com/digitalasset/daml/lf/codegen/CodeGenRunnerTests.scala [hunk 3]
```diff
@@ -21,12 +21,15 @@ class CodeGenRunnerTests extends FlatSpec {
 
   val dummyOutputDir = Files.createTempDirectory("codegen")
 
+  it should "always use JavaBackend, which is currently hardcoded" in {
+    assert(CodeGenRunner.backend eq JavaBackend)
```
[the remark this answers is not in the document]
Comment (leo-da) [reply language-support/java/codegen/src/test/scala/com/digitalasset/daml/lf/codegen/CodeGenRunnerTests.scala]: nah, the test already using assert, no reason to introduce should, plus I quietly hate scalatest Matchers.

[the remark this answers is not in the document]
Comment (S11001001) [reply language-support/java/codegen/src/test/scala/com/digitalasset/daml/lf/codegen/CodeGenRunnerTests.scala]: The problem is the error reporting is terrible without `Matchers`.

[the remark this answers is not in the document]
Comment (leo-da) [reply language-support/java/codegen/src/test/scala/com/digitalasset/daml/lf/codegen/CodeGenRunnerTests.scala]: I know.... but it also does not typecheck at compile time... I choose typecheck
plus it is not my test... I don't want to change the entire thing to use matchers.

[the remark this answers is not in the document]
Comment (S11001001) [reply language-support/java/codegen/src/test/scala/com/digitalasset/daml/lf/codegen/CodeGenRunnerTests.scala]: `eq` isn't typechecked either; I don't think you should change every test to use matchers, just this one.

[the remark this answers is not in the document]
Comment (leo-da) [reply language-support/java/codegen/src/test/scala/com/digitalasset/daml/lf/codegen/CodeGenRunnerTests.scala]: plus the suggested change won't work:
```
class CodeGenRunnerTests extends FlatSpec {
```

[the remark this answers is not in the document]
Comment (S11001001) [reply language-support/java/codegen/src/test/scala/com/digitalasset/daml/lf/codegen/CodeGenRunnerTests.scala]: I have only written the suggestion, not compiled it :)

**Pull request merged**